### ISSUE TITLE
Unicorn closing postgres connection before fork

### ISSUE THREAD

**Issue by dspaeth-faber** [krists/refile-postgres#2]
Hi,

I tested refile postgres with unicorn. There are some issues with the connection. When
used with `preload_app true` DB connections are closed before fork and so this gem doesn't work any more. Is there a way to reinject a new connection?

regards 
Dieter

**Comment by krists** [comment]
Hello @dspaeth-faber 
Thanks for trying out refile-postgres and reporting this issue. I will take a look at it later today. Have not tested it with unicorn before.

**Comment by krists** [comment]
Unfortunately I could not reproduce the error. If you could provide a sample application where it fails, then that would be great.

What you could try is to reset  `Refile` storage configuration in your unicorn's `after_fork` callback.

Have not used unicorn before but config file could look something like this:

``` ruby
timeout 30
preload_app true
worker_processes 2

before_fork do |server, worker|
  ActiveRecord::Base.connection.disconnect! if defined?(ActiveRecord::Base)
end

after_fork do |server, worker|
  ActiveRecord::Base.establish_connection if defined?(ActiveRecord::Base)
  Refile.configure do |config|
    config.store = Refile::Postgres::Backend.new(ActiveRecord::Base.connection.raw_connection)
  end
end

```

**Comment by dspaeth-faber** [comment]
That's right. The after_fork hook is my workaround, but's not nice. Unfortunately I can't provide an example app at the moment. I can only provide my unicorn config:

``` ruby
#
#
# generated through capistrano local changes will be overriden
#
###############################################################

worker_processes 3
timeout 10
preload_app true

listen 8000, :backlog => 256

app_path = "../"
# Fill path to your app
working_path = File.join(app_path, 'current')
working_directory working_path

# feel free to point this anywhere accessible on the filesystem
pid "../tmp/unicorn.pid"

# By default, the Unicorn logger will write to stderr.
# Additionally, ome applications/frameworks log to stderr or stdout,
# so prevent them from going to /dev/null when daemonized here:
stderr_path "#{app_path}/shared/log/unicorn.stderr.log"
stdout_path "#{app_path}/shared/log/unicorn.stdout.log"

before_exec do |server|
  ENV["BUNDLE_GEMFILE"] = "#{working_path}/Gemfile"
end


before_fork do |server, worker|
  Signal.trap 'TERM' do
    #logger.info 'Unicorn master intercepting TERM and sending myself QUIT instead'
    Process.kill 'QUIT', Process.pid
  end

  defined?(ActiveRecord::Base) and
      ActiveRecord::Base.connection.disconnect!

  #http://orchestrate.io/blog/2014/03/26/tip-seamless-restarts-with-unicorn-and-upstart/
  f = File.open("#{server.config[:pid]}.lock", 'w')
  exit unless f.flock(File::LOCK_SH)
end

after_fork do |server, worker|
  Signal.trap 'TERM' do
    #logger.info 'Unicorn worker intercepting TERM and doing nothing. Wait for master to send QUIT'
  end

  defined?(ActiveRecord::Base) and
      ActiveRecord::Base.establish_connection

# with the next line it works, without not
  Refile.configure do |config|
    config.store = Refile::Postgres::Backend.new(ActiveRecord::Base.connection.raw_connection)
  end

  child_pid = server.config[:pid].sub('.pid', ".#{worker.nr}.pid")
  system("echo #{Process.pid} > #{child_pid}")
end

```

Isn't it possible the use connection parameters or lambda instead of a raw connection?

**Comment by krists** [comment]
Yes. It could be done, but it would make configuration a bit more complex.
Maybe there is a more appropriate place for this Refile storage configuration? I will take a look at similar projects and how they avoid `after_fork` use.

**Comment by dspaeth-faber** [comment]
I think the initializer is great, but it need a concept for reconnecting. With a raw connection you can't do this. Or maybe a callback for the case of a connection close.

**Comment by dspaeth-faber** [comment]
Should be fixed in 1e6c796
